fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java

```diff
@@ -0,0 +1,184 @@
+
+            if (found.isPresent()) {
+                if (found.get() instanceof NamedExpression) {
+                    substitution.put(expression, ((NamedExpression) found.get()).toSlot());
+                    if (!isFoundInOutputExpressions) {
+                        newOutputSlots.add(((NamedExpression) found.get()).toSlot());
+                    }
+                } else {
+                    // This expression was found in group-by expressions. E.g. group by (a + 1) having (a + 1)
+                    generateAliasForNewOutputSlots(expression);
+                }
+            } else {
+                if (expression instanceof SlotReference) {
+                    throw new AnalysisException(expression.toSql() + " in having clause should be grouped by.");
+                } else if (expression instanceof AggregateFunction) {
+                    if (checkWhetherNestedAggregateFunctionsExist((AggregateFunction) expression)) {
+                        throw new AnalysisException("Aggregate functions in having clause can't be nested: "
+                                + expression.toSql() + ".");
+                    }
+                    generateAliasForNewOutputSlots(expression);
+                } else {
+                    for (Expression child : expression.children()) {
+                        resolve(child);
+                    }
+                }
+            }
+        }
+
+        private Pair<Optional<Expression>, Boolean> lookUp(Expression expression) {
+            Optional<Expression> found = outputExpressions.stream()
+                    .filter(source -> isEquivalent(source, expression))
+                    .map(source -> (Expression) source)
+                    .findFirst();
+            if (found.isPresent()) {
+                return Pair.of(found, true);
+            }
+
+            found = groupByExpressions.stream().filter(source -> isEquivalent(source, expression)).findFirst();
+            return Pair.of(found, false);
+        }
+
+        private boolean isEquivalent(Expression source, Expression expression) {
+            if (source.equals(expression)) {
+                return true;
+            } else if (source instanceof Alias) {
+                Alias alias = (Alias) source;
+                return alias.toSlot().equals(expression) || alias.child().equals(expression);
+            }
+            return false;
+        }
+
+        private boolean checkWhetherNestedAggregateFunctionsExist(AggregateFunction function) {
+            Predicate<Expression> hasAggregateFunctions = new Predicate<Expression>() {
+                @Override
+                public boolean test(Expression expression) {
+                    return expression instanceof AggregateFunction
+                            || expression.children().stream().anyMatch(this);
+                }
+            };
+            return function.children().stream().anyMatch(hasAggregateFunctions);
```

Comment:
could we just use?
```
function.children().stream().anyMatch(c -> c.anyMatch(AggregateFunction.class::isInstance));
```

Reply:
We need recursion here otherwise we can't cover the case `sum(a1 + a2 + sum(a2))` which is parsed as `new Sum(new Add(new Add(a1, a2), new Sum(a2)))`

Reply:
Ok, the modification also works. I use this way.